ktor-http/ktor-http-cio/src/io/ktor/http/cio/internals/WeakTimeoutQueue.kt

```diff
@@ -74,34 +74,29 @@ class WeakTimeoutQueue(
     suspend fun <T> withTimeout(block: suspend CoroutineScope.() -> T): T {
         return suspendCoroutineUninterceptedOrReturn { rawContinuation ->
             val continuation = rawContinuation.intercepted()
+
             val wrapped = WeakTimeoutCoroutine(continuation.context, continuation)
             val handle = register(wrapped)
-            wrapped.disposeOnCompletion(handle)
+            wrapped.invokeOnCompletion(handle)
 
             val result = try {
-                if (wrapped.isCancelled) throw wrapped.getCancellationException()
-                block.startCoroutineUninterceptedOrReturn(receiver = wrapped, completion = wrapped)
+                if (wrapped.isCancelled) COROUTINE_SUSPENDED
```

Comment:
`if (wrapped.isCancelled) COROUTINE_SUSPENDED`
But cancellation exception should be thrown instead immediately?

Reply:
If it is already cancelled then it was already resumed this is why we should suspend here. However it could be improved for sure, just not priority for now